core/include/tcb.h

```diff
@@ -38,34 +39,38 @@
  * @brief Blocked states.
  * @{
  */
-#define STATUS_STOPPED          0               /**< has terminated                     */
-#define STATUS_SLEEPING         1               /**< sleeping                           */
-#define STATUS_MUTEX_BLOCKED    2               /**< waiting for a locked mutex         */
-#define STATUS_RECEIVE_BLOCKED  3               /**< waiting for a message              */
-#define STATUS_SEND_BLOCKED     4               /**< waiting for message to be delivered*/
-#define STATUS_REPLY_BLOCKED    5               /**< waiting for a message response     */
+#define STATUS_STOPPED              0   /**< has terminated                     */
+#define STATUS_SLEEPING             1   /**< sleeping                           */
+#define STATUS_MUTEX_BLOCKED        2   /**< waiting for a locked mutex         */
+#define STATUS_RECEIVE_BLOCKED      3   /**< waiting for a message              */
+#define STATUS_SEND_BLOCKED         4   /**< waiting for message to be delivered*/
+#define STATUS_REPLY_BLOCKED        5   /**< waiting for a message response     */
+#define STATUS_FLAG_BLOCKED_ANY     6   /**< waiting for any flag from flag_mask*/
+#define STATUS_FLAG_BLOCKED_ALL     7   /**< waiting for all flags in flag_mask */
 /** @} */
 
 /**
  * @brief These have to be on a run queue.
  * @{*/
 #define STATUS_ON_RUNQUEUE      STATUS_RUNNING  /**< to check if on run queue:
-                                                 `st >= STATUS_ON_RUNQUEUE`             */
-#define STATUS_RUNNING          6               /**< currently running                  */
-#define STATUS_PENDING          7               /**< waiting to be scheduled to run     */
+                                                 `st >= STATUS_ON_RUNQUEUE`         */
+#define STATUS_RUNNING          9               /**< currently running              */
+#define STATUS_PENDING          10              /**< waiting to be scheduled to run */
 /** @} */
 /** @} */
 
 /**
  * @brief @c tcb_t holds thread's context data.
  */
-typedef struct tcb_t {
+struct tcb {
     char *sp;                   /**< thread's stack pointer         */
     uint16_t status;            /**< thread's status                */
 
     kernel_pid_t pid;           /**< thread's process id            */
     uint16_t priority;          /**< thread's priority              */
 
+    thread_flags_t flags;       /**< currently set flags            */
```

Comment:
It could. But I've also changed msg.c to use thread flags internally, which makes some things a lot nicer.

e.g., currently in xtimer_receive_msg_timeout, there's a race between receiving the message and deleting the timer, and if the thread has a message queue, that might lead to the timer message to be queued right after the actual received message.

This PR's xtimer_receive_msg_timeout() implementation uses a thread flag to pass the timeout information, solving the race problem.

The pseudomodule basically only disables the thread flags user API.